File: java/BaseClass.java
public class BaseClass {
    protected String baseVar;

    public BaseClass() {
        this.baseVar = "I'm a base class variable";
    }

    public void baseMethod() {
        System.out.println("This is a method from BaseClass");
        System.out.println("Base var: " + baseVar);
    }

    // Demonstrate generic method
    public <T> void printArray(T[] array) {
        for (T element : array) {
            System.out.print(element + " ");
        }
        System.out.println();
    }
}

File: ruby/base_class.rb
class BaseClass
  @@class_variable = "I'm a class variable in BaseClass"

  def initialize
    @base_var = "I'm a base class variable"
  end

  def base_method
    puts "This is a method from BaseClass"
    puts "Base var: #{@base_var}"
  end

  def self.class_variable
    @@class_variable
  end

  # Demonstrate metaprogramming
  def self.define_dynamic_method(method_name)
    define_method(method_name) do
      puts "This is a dynamically defined method: #{method_name}"
    end
  end
end